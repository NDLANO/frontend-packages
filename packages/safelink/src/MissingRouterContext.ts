/**
 * Copyright (c) 2019-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { createContext } from "react";

const MissingRouterContext = createContext(false);

export default MissingRouterContext;
