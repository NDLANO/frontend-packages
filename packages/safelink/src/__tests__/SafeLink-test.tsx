/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';
import { StaticRouter } from 'react-router';
import Safelink, { isOldNdlaLink } from '../SafeLink';
import MissingRouterContext from '../MissingRouterContext';

test('SafeLink renderers Link correctly if router context is present', () => {
  const component = renderer.create(
    <StaticRouter location="foo" context={{}}>
      <div>
        <Safelink to="/my/path">Snapshot should contain onClick</Safelink>,
      </div>
    </StaticRouter>,
  );

  expect(component.toJSON()).toMatchSnapshot();
  // @ts-ignore
  expect(component.toJSON()!.children![0].props.onClick).toBeInstanceOf(
    Function,
  );
});

test('SafeLink defaults to normal link if to prop is an external link', () => {
  const component = renderer.create(
    <StaticRouter location="foo" context={{}}>
      <div>
        <Safelink to="https://example.com">
          Snapshot should not contain onClick
        </Safelink>
        ,
      </div>
    </StaticRouter>,
  );

  expect(component.toJSON()).toMatchSnapshot();
  // @ts-ignore
  expect(component.toJSON()!.children![0].props.onClick).toBeUndefined();
});

test('SafeLink defaults to normal link if to prop is an old ndla link', () => {
  const component = renderer.create(
    <StaticRouter location="foo" context={{}}>
      <div>
        <Safelink to="/nb/node/54">
          Snapshot should not contain onClick
        </Safelink>
        ,
      </div>
    </StaticRouter>,
  );

  // @ts-ignore
  expect(component.toJSON()!.children![0].props.onClick).toBeUndefined();
});

test('SafeLink renderers normal link correctly when router context is not present', () => {
  const component = renderer.create(
    <MissingRouterContext.Provider value={true}>
      <Safelink to="/my/path">No router context</Safelink>
    </MissingRouterContext.Provider>,
  );

  expect(component.toJSON()).toMatchSnapshot();
});

test('isOldNdlaLink checks', () => {
  expect(isOldNdlaLink('/nb/node/12')).toBe(true);
  expect(isOldNdlaLink('/nn/node/12?fag=12')).toBe(true);
  expect(isOldNdlaLink('/en/node/12ssjkdlf')).toBe(true);
  expect(isOldNdlaLink('/nb/nde/12')).toBe(false);
  expect(isOldNdlaLink('/subjects')).toBe(false);
  expect(isOldNdlaLink('/sanodesd43/')).toBe(false);
  expect(isOldNdlaLink({ pathname: '/sanodesd43/' })).toBe(false);
});
