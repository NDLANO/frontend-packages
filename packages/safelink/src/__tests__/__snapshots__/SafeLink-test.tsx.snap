// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SafeLink defaults to normal link if to prop is an external link 1`] = `
<div>
  <div>
    <a
      href="https://example.com"
    >
      Snapshot should not contain onClick
    </a>
    ,
  </div>
</div>
`;

exports[`SafeLink defaults to normal link if to prop is an old ndla link 1`] = `
<div>
  <div>
    <a
      href="/nb/node/54"
    >
      Snapshot should not contain onClick
    </a>
    ,
  </div>
</div>
`;

exports[`SafeLink renderers Link correctly if router context is present 1`] = `
<div>
  <div>
    <a
      href="/my/path"
      tabindex="0"
    >
      Snapshot should contain onClick
    </a>
    ,
  </div>
</div>
`;

exports[`SafeLink renderers normal link correctly when router context is not present 1`] = `
<div>
  <a
    href="/my/path"
  >
    No router context
  </a>
</div>
`;
